BGL/include/CGAL/boost/graph/helpers.h

```diff
@@ -355,29 +356,207 @@ bool is_valid_face_descriptor( typename boost::graph_traits<FaceGraph>::face_des
+  typedef typename boost::graph_traits<PolygonMesh>::halfedge_iterator     halfedge_const_iterator;
+  typedef typename boost::graph_traits<PolygonMesh>::face_iterator         face_const_iterator;
+  typedef typename boost::graph_traits<PolygonMesh>::vertices_size_type    vertex_size_type;
+  typedef typename boost::graph_traits<PolygonMesh>::halfedges_size_type   halfedges_size_type;
+  typedef typename boost::graph_traits<PolygonMesh>::faces_size_type       faces_size_type;
+  Verbose_ostream verr(verb);
+  int num_v(std::distance(boost::begin(vertices(g)), boost::end(vertices(g)))),
+  num_h(std::distance(boost::begin(halfedges(g)), boost::end(halfedges(g)))),
+  num_f(std::distance(boost::begin(faces(g)), boost::end(faces(g))));
+  bool valid = ( 1 != (num_h& 1));
+  if ( ! valid)
+    verr << "number of halfedges is odd." << std::endl;
+
+  // All halfedges.
+
+  halfedges_size_type  n = 0;
+  halfedges_size_type nb = 0;
+  BOOST_FOREACH(halfedge_descriptor begin, halfedges(g)) {
+    if(!valid)
+      break;
+    verr << "halfedge " << n << std::endl;
+    if ( is_border(begin, g))
+      verr << "    is border halfedge" << std::endl;
+    // Pointer integrity.
+    valid = valid && ( next(begin, g) != boost::graph_traits<PolygonMesh>::null_halfedge());
+    valid = valid && ( opposite(begin, g) != boost::graph_traits<PolygonMesh>::null_halfedge());
+    if ( ! valid) {
+      verr << "    pointer integrity corrupted (ptr==0)."
+           << std::endl;
+      break;
+    }
+    // opposite integrity.
+    valid = valid && ( opposite(begin, g) != begin);
+    valid = valid && ( opposite(opposite(begin, g), g) == begin);
+    if ( ! valid) {
+      verr << "    opposite pointer integrity corrupted."
+           << std::endl;
+      break;
+    }
+    // previous integrity.
+    valid = valid && ( prev(next(begin, g), g) == begin);
+    if ( ! valid) {
+      verr << "    previous pointer integrity corrupted."
+           << std::endl;
+      break;
+    }
+    // vertex integrity.
+    valid = valid && ( target(begin, g) != boost::graph_traits<PolygonMesh>::null_vertex());
+    if ( ! valid) {
+      verr << "    vertex pointer integrity corrupted."
+           << std::endl;
+      break;
+    }
+    valid = valid && ( target(begin, g) ==
+                       target(opposite(next(begin, g), g), g));
+    if ( ! valid) {
+      verr << "    vertex pointer integrity2 corrupted."
+           << std::endl;
+      break;
     }
+    // face integrity.
+
+    valid = valid && ( face(begin, g) == face(next(begin, g), g));
+    if ( ! valid) {
+      verr << "    face pointer integrity2 corrupted."
+           << std::endl;
+      break;
+    }
+
+    ++n;
+    if ( is_border(begin, g))
+      ++nb;
   }
-  BOOST_FOREACH(halfedge_descriptor h, halfedges(g)){
-    if(! is_valid_halfedge_descriptor(h,g)){
-      return false;
+  verr << "summe border halfedges (2*nb) = " << 2 * nb << std::endl;
```

Comment:
sum of ?

Reply:
Probably. It is a copy/paste from Polyhedron_3::is_valid(). I guess I should change this one too